### PR TITLE
[JUJU-3333] Users removal triggers the removal of model access

### PR DESCRIPTION
**Pull request by juanmanuel-tirado**
After enabling user names to be reused (#15266 ) it has been observed that for those recreated users that where previously granted with access permissions to other models, these permissions are automatically recreated. This is a misbehaviour that is corrected in this PR by removing all the permissions a user is granted when the user is removed from the controller. This makes possible to have a clear list of permissions when a user is recreated avoiding this unexpected behaviour.

## Checklist

*If an item is not applicable, use `~strikethrough~`.*

~- [ ] Code style: imports ordered, good names, simple structure, etc~
~- [ ] Comments saying why design decisions were made~
- [x] Go unit tests, with comments saying what you're testing
~- [ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
~- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Try to replicate an scenario for a recreated user who was granted some permissions.

```sh
juju add-user mrx
User "mrx" added
Please send this command to mrx:
    juju register MHsTA21yeDBFDC5bZmQ0MjpjZmExOmFkZjc6NjI3NDoyMTY6M2VmZjpmZTE1OjZiYzldOjE3MDcwExMxMC4yMjAuMTguMjIyOjE3MDcwBCC9TIZM91FQZPJ2bjnICsB4rcyUlcpmPXrxgpQpxJx5NRMJbG9jYWx0ZXN0EwAA

"mrx" has not been granted access to any models. You can use "juju grant" to grant access.
```
Grant user access
```sh
juju add-model test
juju grant mrx admin test
juju users test
Name    Display name  Access  Last connection
admin*  admin         admin   36 minutes ago
mrx                   admin   never connected
```
Destroy the user and recreate
```sh
juju remove-user mrx --yes
users test
Name    Display name  Access  Last connection
admin*  admin         admin   37 minutes ago

juju add-user mrx
User "mrx" added
Please send this command to mrx:
    juju register MHsTA21yeDBFDC5bZmQ0MjpjZmExOmFkZjc6NjI3NDoyMTY6M2VmZjpmZTE1OjZiYzldOjE3MDcwExMxMC4yMjAuMTguMjIyOjE3MDcwBCC9TIZM91FQZPJ2bjnICsB4rcyUlcpmPXrxgpQpxJx5NRMJbG9jYWx0ZXN0EwAA

"mrx" has not been granted access to any models. You can use "juju grant" to grant access.
```

Now if you check the users for model `test`, no permissions are set
```sh
users test
Name    Display name  Access  Last connection
admin*  admin         admin   37 minutes ago
```

### CODE REVIEW

File: state/user.go [hunk 1]
```diff
@@ -186,7 +184,7 @@ func (st *State) updateExistingUser(u *User, displayName, password, creator stri
 		Assert: txn.DocExists,
 		Update: updateUser,
 	}}
-	updateUserOps = append(updateUserOps, removeControllerOps...)
+	// updateUserOps = append(updateUserOps, removeControllerOps...)
```
Comment (nvinuesa) [inline state/user.go:187]: this can be removed

**Review by nvinuesa — APPROVED** [review]
LGTM, only a minor comment. QA passes

File: state/user.go (hunk 1 above)
Comment (juanmanuel-tirado) [reply state/user.go]: Removed. Thx.

File: state/user.go [hunk 2]
```diff
@@ -215,6 +212,28 @@ func (st *State) RemoveUser(tag names.UserTag) error {
 		return nil
 	}
 
+	// remove the access to all the models and the current controller
+	// first query all the models for this user
+	modelsSummary, err := st.ModelSummariesForUser(tag, true)
```
Comment (hpidcock) [inline state/user.go:217]: This is not how you would iterate across all the model users.

File: state/user.go [hunk 3]
```diff
@@ -215,6 +212,28 @@ func (st *State) RemoveUser(tag names.UserTag) error {
...
+	if err != nil {
+		return err
+	}
+
+	for _, summary := range modelsSummary {
+		// remove all the access permissions
+		removeModelOps := removeModelUserOps(summary.UUID, tag)
```
Comment (hpidcock) [inline state/user.go:224]: removeModelUserOps can't be run like this. You'll need to manually write the ops for this, as the removeModelUserOps are meant to be run from the state opened for a specific model, so you'll need to handle model prefixed ids manually. See ensureModelUUID and modelStateCollection and GetCollection.

File: state/user.go [hunk 4]
```diff
@@ -215,6 +212,28 @@ func (st *State) RemoveUser(tag names.UserTag) error {
...
+		if err := st.db().RunTransactionFor(summary.UUID, removeModelOps); err != nil {
+			logger.Errorf("impossible to remove user from models", err)
+			return err
+		}
+	}
+
+	// remove the user from the user controllers
+	if err := st.removeControllerUser(tag); err != nil {
```
Comment (hpidcock) [inline state/user.go:232]: You should actually be getting this user from state first to check it still exists.

File: state/user.go [hunk 5]
```diff
@@ -235,13 +254,14 @@ func (st *State) RemoveUser(tag names.UserTag) error {
 				return nil, errors.Trace(err)
 			}
 		}
```
Comment (hpidcock) [inline state/user.go:256]: You should be building all of those ops and txns above into this buildTxn function so they can be done together.

**Comment by hpidcock** [comment]
Please address this comment too. https://github.com/juju/juju/pull/15279/commits/c42e57ebc1d0c264f5bca0207b4ad3866a65ed1d#r1151429992

**Pull request merged**